packages/flutter/lib/src/painting/text_span.dart

```diff
@@ -196,78 +202,113 @@ class TextSpan extends DiagnosticableTree {
+      return this;
+    }
+    offset.value = endOffset;
+    return null;
+  }
 
   /// Flattens the [TextSpan] tree into a single string.
   ///
   /// Styles are not honored in this process. If `includeSemanticsLabels` is
   /// true, then the text returned will include the [semanticsLabel]s instead of
   /// the text contents when they are present.
-  String toPlainText({bool includeSemanticsLabels = true}) {
-    assert(debugAssertIsValid());
+  ///
+  /// When [includePlaceholders] is true, [PlaceholderSpan]s in the tree will be
+  /// represented as a 0xFFFC 'object replacement character'.
+  @override
+  String toPlainText({bool includeSemanticsLabels = true, bool includePlaceholders = true}) {
     final StringBuffer buffer = StringBuffer();
-    visitTextSpan((TextSpan span) {
-      if (span.semanticsLabel != null && includeSemanticsLabels) {
-        buffer.write(span.semanticsLabel);
-      } else {
-        buffer.write(span.text);
-      }
-      return true;
-    });
+    computeToPlainText(buffer, includeSemanticsLabels: includeSemanticsLabels, includePlaceholders: includePlaceholders);
     return buffer.toString();
   }
+  @override
+  void computeToPlainText(StringBuffer buffer, {bool includeSemanticsLabels = true, bool includePlaceholders = true}) {
+    assert(debugAssertIsValid());
+    if (semanticsLabel != null && includeSemanticsLabels) {
+      buffer.write(semanticsLabel);
+    } else if (text != null) {
+      buffer.write(text);
+    }
+    if (children != null) {
+      for (InlineSpan child in children) {
+        child.computeToPlainText(buffer,
+          includeSemanticsLabels: includeSemanticsLabels,
+          includePlaceholders: includePlaceholders,
+        );
+      }
+    }
+  }
 
   /// Returns the UTF-16 code unit at the given index in the flattened string.
   ///
+  /// This only accounts for the [TextSpan.text] values and ignores [PlaceholderSpans].
+  ///
   /// Returns null if the index is out of bounds.
+  @override
   int codeUnitAt(int index) {
     if (index < 0)
       return null;
-    int offset = 0;
-    int result;
-    visitTextSpan((TextSpan span) {
-      if (index - offset < span.text.length) {
-        result = span.text.codeUnitAt(index - offset);
-        return false;
-      }
-      offset += span.text.length;
-      return true;
+    final TrackingInt offset = TrackingInt();
+    final TrackingInt result = TrackingInt(null);
+    visitChildren((InlineSpan span) {
+      return span.codeUnitAtVisitor(TrackingInt(index), offset, result);
+
     });
-    return result;
+    return result.value;
+  }
+  @override
+  bool codeUnitAtVisitor(TrackingInt index, TrackingInt offset, TrackingInt result) {
+    if (text == null) {
+      return true;
+    }
+    if (index.value - offset.value < text.length) {
+      result.value = text.codeUnitAt(index.value - offset.value);
+      return false;
+    }
+    offset.value += text.length;
```

Comment:
given how you use the TrackingInt for the value here, i'd probably have something like:

```dart
class Accumulator {
  Accumulator([this._value = 0]);
  int get value => _value;
  int _value;
  void increment(int addend) {
    assert(addend >= 0);
    value += addend;
  }
}
```